android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt

```diff
@@ -315,3 +325,48 @@ fun String.extractLogicalIdUuid() = this.substringAfter("/").substringBefore("/"
 fun Resource.addTags(tags: List<Coding>) {
   tags.forEach { this.meta.addTag(it) }
 }
+
+suspend fun Task.updateDependentTaskDueDate(defaultRepository: DefaultRepository): Task {
+  return apply {
+    val isUpcoming = status == Task.TaskStatus.REQUESTED
+    val isOverdue =
+      (status == Task.TaskStatus.READY || status == Task.TaskStatus.INPROGRESS) &&
+        executionPeriod.end.before(Date())
+    val isDue = status == Task.TaskStatus.INPROGRESS
+    if (!(isUpcoming || isOverdue || isDue) || partOf.isEmpty()) {
+      return this
+    }
+    partOf.forEach { task ->
+      val dependentTask = defaultRepository.loadResource(task.reference as Reference) as? Task
+      dependentTask?.output?.forEach { dependentTaskOutputValue ->
+        val encounterResource =
+          defaultRepository.loadResource(
+            readValueFromJsonString(dependentTaskOutputValue.valueToString(), "reference")
```

Comment:
think this can use JsonExtension
```suggestion
            dependentTaskOutputValue.valueToString().decodeJson()["reference"]!!.jsonPrimitive.content
```

Reply:
`decodeJson`  will only work for serializable kotlin classes and we have not implemented any kotlin serialization for FHIR resources.

But I agree with @pld We do not need this `readValueFromJsonString` extension. Use existing extensions if necessary.

Reply:
The loadResource accepts a generic Resource type e.g. `loadResource<Encounter>`, use this instead of casting to Encounter resource. Will the resource type always be Encounter? if you are not sure, it will be safe to use ` loadResource(resourceId: String, resourceType: ResourceType)` instead so you can provide the inferred resource type.